gen/spec.go

```diff
@@ -199,6 +224,181 @@ type traversal interface {
 	gStruct(*Struct)
 }
 
+type declarer interface {
+	declareOnce(p printer, name, typ string)
+}
+
+type assigner interface {
+	assignAndCheck(name, base string)
+	nextTypeAndCheck(name string)
+	skipAndCheck()
+}
+
+type fuser interface {
+	Fuse([]byte)
+}
+
+type traversalAssigner interface {
+	traversal
+	assigner
+	declarer
+}
+
+type traversalFuser interface {
+	traversal
+	fuser
+}
+
+func genStructFieldsSerializer(t traversalFuser, p printer, fields []StructField) {
+	data := msgp.AppendMapHeader(nil, uint32(len(fields)))
+	p.printf("\n// map header, size %d", len(fields))
+	t.Fuse(data)
+	for _, f := range fields {
+		if !p.ok() {
+			return
+		}
+
+		data, p.err = msgp.AppendIntf(nil, f.FieldTag)
+		p.printf(
+			"\n// [field %q] write label `%v` as msgp.%s",
+			f.FieldName, f.FieldTag, msgp.NextType(data),
+		)
+		t.Fuse(data)
+
+		next(t, f.FieldElem)
+	}
+}
+
+func genStructFieldsParser(t traversalAssigner, p printer, fields []StructField) {
+	const (
+		fieldBytes = "field_b"
```

Comment:
Using `snake_case` will trip up tools like `golint`. It would be friendlier to use something more idiomatic.